core/src/main/java/com/linecorp/armeria/internal/common/Http1ObjectEncoder.java

```diff
@@ -112,13 +115,46 @@ protected final ChannelFuture writeNonInformationalHeaders(int id, HttpObject co
         } else {
             f = write(id, converted, false);
             if (endStream) {
-                f = write(id, LastHttpContent.EMPTY_LAST_CONTENT, true);
+                f = combine(f, write(id, LastHttpContent.EMPTY_LAST_CONTENT, true));
             }
         }
         ch.flush();
         return f;
     }
 
+    private ChannelPromise combine(ChannelFuture first, ChannelFuture second) {
+        final ChannelPromise promise = channel().newPromise();
+        final FutureListener<Void> listener =
+                new FutureListener<Void>() {
+                    private int remaining = 2;
+
+                    @Override
+                    public void operationComplete(Future<Void> ignore) throws Exception {
+                        remaining--;
+                        if (remaining == 0) {
+                            final Throwable firstCause = first.cause();
+                            final Throwable secondCause = second.cause();
+                            if (firstCause != null) {
+                                if (Flags.verboseExceptionSampler().isSampled(firstCause.getClass()) &&
+                                    secondCause != null) {
```

Comment:
Given this doesn't give any performance benefit, I'd suggest reverting this:

```suggestion
                                if (secondCause != null) {
```